File: AppAndroid/app/src/main/java/com/example/mobiletiltmouse/RemoteAccess.kt
package com.example.mobiletiltmouse

import android.content.Context

/**
 * Coordinates remote control functionalities by managing network discovery,
 * device pairing, connection handling, mouse action execution, and network monitoring.
 *
 * @param context The application context used to initialize necessary system services.
 */
class RemoteAccess(val context: Context, userSettings: UserSettings) {
    var networkMonitor = NetworkMonitor(context)
    var mouseActions   = MouseActions(context)
    var pairing        = Pairing(mouseActions, userSettings)
    private var connection: Connection? = null
    private var nwBrowser: NetworkBrowser? = null

    init {
        connection = Connection(context, this, pairing)
        nwBrowser = NetworkBrowser(context, connection, this)
    }

    fun startRemoteAccess() {
        nwBrowser?.startBrowsing()
        networkMonitor.startNetworkMonitor()
    }

    fun stopRemoteAccess() {
        connection?.stopConnection()
        networkMonitor.stopNetworkMonitor()
    }

    fun restartNetwork() {
        connection?.stopConnection()
        nwBrowser?.stopBrowsing()
        Thread.sleep(1000)
        // NsdManger's resolve listener might not be released, therefore a new instance is created
        nwBrowser = NetworkBrowser(context, connection, this)
        nwBrowser?.startBrowsing()
    }
}

File: AppiOS/MobileTiltMouse/RemoteAccess.swift
import Foundation

/// Coordinates remote control functionalities by managing network discovery,
/// device pairing, connection handling, mouse action execution, and network monitoring.
///
/// Start sequence:
/// 1. Search server in network - NetworkBrowser()
/// 2. Connect to server - Connection()
/// 3. Check Pairing with server - Pairing()
/// 4. Start mouse control - MouseActions()
/// Each class calls the start of the next step (and might pass parameters).
///
/// Network monitoring starts automatically on initialization to inform the user about network interface availability.
///
/// - Parameters:
///     - errAlert: For showing error alert
///     - networkStatus: Status of network connectivity
///     - pairingStatus: Status of pairing process
///
class RemoteAccess {
    var connection: Connection?
    var nwBrowser: NetworkBrowser?
    var mouseAction: MouseActions?
    var pairing: Pairing?
    
    init(errAlert: ErrorAlert, networkStatus: NWStatus, pairingStatus: PairingStatus) {
        mouseAction = MouseActions()
        pairing = Pairing(mouseAction: mouseAction, pairingStatus: pairingStatus, networkStatus: networkStatus)
        connection = Connection(errAlert: errAlert, networkStatus: networkStatus, pairing: pairing)
        nwBrowser = NetworkBrowser(networkStatus: networkStatus, connection: connection)
        
        NetworkMonitor(networkStatus: networkStatus).startMonitoring()
    }
    
    func startRemoteAccess() {
        nwBrowser?.startBrowsing()
    }
    
    func stopRemoteAccess() {
        mouseAction?.stopMotionUpdate()
        connection?.stopConnection()
    }
}
